cirq/ops/projectors.py

```diff
@@ -0,0 +1,103 @@
+                 qid_shape: Tuple[int, ...] = (2,),
+                 enforce_orthonormal_basis: bool = False):
+        """
+        Args:
+            projection_basis: a (p, 2**num_qubits) matrix that lists the
+                projection vectors, where p is the dimension of the subspace
+                we're projecting on. If you project onto a single vector, then
+                p = 1 and thus the matrix reduces to a single row.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+            enforce_orthonormal_basis: Whether to enfore the input basis to be
+                orthogonal.
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        projection_array = np.asarray(projection_basis)
+
+        if len(projection_array.shape) != 2:
+            raise ValueError('The input projection_basis must be a 2D array')
+
+        if enforce_orthonormal_basis:
+            B = projection_array @ projection_array.T.conj()
+            if not np.allclose(B, np.eye(projection_array.shape[0])):
+                raise ValueError('The basis must be orthogonal')
+
+        if np.linalg.matrix_rank(projection_array) < projection_array.shape[0]:
+            raise ValueError('Vectors in basis must be linearly independent')
+
+        if np.prod(qid_shape) != projection_array.shape[1]:
+            raise ValueError(
+                "Invalid shape " +
+                f"{np.array(projection_array).shape} for qid_shape {qid_shape}")
+        self._projection_basis = projection_array
+        self._qid_shape = qid_shape
+
+    def _projection_basis_(self) -> np.ndarray:
+        return self._projection_basis
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _has_unitary_(self) -> bool:
+        return False
+
+    def _is_parameterized_(self) -> bool:
+        return False
+
+    def _channel_(self) -> Iterable[np.ndarray]:
+        A = self._projection_basis
+        pseudoinverse = A.T.conj() @ np.linalg.inv(A @ A.T.conj())
+        return (pseudoinverse @ A,)
+
+    def _has_channel_(self) -> bool:
+        return True
+
+    def __repr__(self) -> str:
+        return ("cirq.Projector(projection_basis=" +
+                f"{self._projection_basis.tolist()})," +
+                f"qid_shape={self._qid_shape})")
+
+    def _circuit_diagram_info_(self,
+                               args: 'protocols.CircuitDiagramInfoArgs') -> str:
+        with np.printoptions(precision=args.precision):
+            return (f"Proj({self._projection_basis.tolist()})")
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return {
+            'cirq_type': self.__class__.__name__,
+            'projection_basis': self._projection_basis.tolist(),
+            'qid_shape': self._qid_shape,
+        }
+
+    def _value_equality_values_(self) -> Any:
+        return self._projection_basis.tolist(), self._qid_shape
```

Comment:
This needs to be hashable

Reply:
Added a test that revealed the failure, then modified the code to make it pass.